File: PatchCore/src/main/cpp/patchcore/module/Module.cpp
/*
 * PatchCore — Modular Synthesizer Engine
 * Copyright (c) 2025 Evgenii Petrov
 *
 * This file is part of PatchCore.
 *
 * PatchCore is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * PatchCore is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU AGPL License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 *
 * Commercial licensing available: contact sillydevices@gmail.com
 */

#include "patchcore/module/Module.hpp"
#include "patchcore/module/PolyModule.hpp"
#include "patchcore/module/PolyProxyModule.hpp"

Module::Module(std::string name, int sampleRate) : sampleRate(sampleRate), name(name) {};

void Module::processBlock() {
    for (int sampleIndex = 0; sampleIndex < PATCHCORE_BLOCK_SIZE; ++sampleIndex) {
        processSample(sampleIndex);
    }
}


const std::string& Module::getModuleName() const {
    return name;
}


const std::unordered_map<std::string, ModuleInput *>& Module::getModuleInputs() const {
    return inputs;
}

ModuleInput *Module::getModuleInput(const std::string& moduleName) {
    return inputs[moduleName];
}


const std::unordered_map<std::string, ModuleOutput *>& Module::getModuleOutputs() const {
    return outputs;
}

ModuleOutput *Module::getModuleOutput(const std::string& outputName) {
    return outputs[outputName];
}


const std::unordered_map<std::string, UserInput *>& Module::getUserInputs() const {
    return userInputs;
}

UserInput *Module::getUserInput(const std::string& inputName) {
    auto input = userInputs[inputName];
    return input;
}

void Module::onStartBlock(const BlockContext& context) {
    for (auto &input : uniqueInputs) {
        input->prepareBlock(context);
    }

    for (auto &output : uniqueOutputs) {
        output->prepareBlock(context);
    }

    for (auto &input : interpolatedInputs) {
        input->prepareBlock(context);
    }
}

int Module::getSampleRate() const {
    return sampleRate;
}

void Module::onProxyOutputCreated(ModuleOutput* output, const std::string& withName) {
    _hasProxyOutputs = true;
}

bool Module::hasProxyOutputs() const {
    return _hasProxyOutputs;
}

const std::vector<FloatUserInput *> &Module::getInterpolatedInputs() {
    return interpolatedInputs;
}

std::unique_ptr<PolyProxyModule> Module::createPolyModuleProxy(PolyModule* polyModule) const {
    return std::make_unique<PolyProxyModule>(this, polyModule);
}

File: PatchCoreMultiplatform/src/commonMain/kotlin/com/sillydevices/patchcore/module/Module.kt
/*
 * PatchCore — Modular Synthesizer Engine
 * Copyright (c) 2025 Evgenii Petrov
 *
 * This file is part of PatchCore.
 *
 * PatchCore is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * PatchCore is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU AGPL License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 *
 * Commercial licensing available: contact sillydevices@gmail.com
 */

package com.sillydevices.patchcore.module

import com.sillydevices.patchcore.context.EmptyModuleContext
import com.sillydevices.patchcore.context.ModuleContext
import com.sillydevices.patchcore.context.PatchModuleContext
import com.sillydevices.patchcore.internal.pointers.ModulePointer
import com.sillydevices.patchcore.module.io.input.ModuleInput
import com.sillydevices.patchcore.module.io.output.ModuleOutput
import com.sillydevices.patchcore.module.io.user.UserInput

abstract class Module(val name: String) {

    protected val registeredInputs = mutableListOf<ModuleInput>()
    protected val registeredOutputs = mutableListOf<ModuleOutput>()
    protected val registeredUserInputs = mutableListOf<UserInput>()

    protected fun registerInput(name: String): ModuleInput {
        val input = ModuleInput(this.name, name)
        registerInput(input)
        return input
    }

    private fun <T:ModuleInput> registerInput(input: T): T {
        registeredInputs.add(input)
        return input
    }

    protected fun registerOutput(name: String): ModuleOutput {
        val output = ModuleOutput(this.name, name)
        registerOutput(output)
        return output
    }

    private fun <T:ModuleOutput> registerOutput(output: T): T {
        registeredOutputs.add(output)
        return output
    }

    protected fun <T: UserInput> registerUserInput(userInput: T): T {
        registeredUserInputs.add(userInput)
        return userInput
    }

    protected var moduleContext: ModuleContext = EmptyModuleContext
        private set

    open fun applyContext(context: ModuleContext) {
        if (moduleContext !is EmptyModuleContext) {
            throw IllegalStateException("Module context is already set. Cannot apply context again.")
        }
        this.moduleContext = context
    }

    //Context and so on
    abstract fun createFrom(context: PatchModuleContext)

    open val pointer: ModulePointer
        get() = moduleContext.getPointer()
}











